File: 5/fcfs.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <deque>
#include <sstream>
#include <string>
#include <algorithm>
#include <cstring>
#include <queue>
#include <limits.h>

using namespace std;

class fc{
    public:
    int ready;
    deque<int> d;
    int fin = 0;
    bool rd = true;
    bool cntflag = false;
};

struct Num
{
	int index, num;
	Num(int index, int num) : index(index), num(num){}
};

struct MyComparator
{
	bool operator()(const Num& o1, const Num& o2)
	{
		
			if (o1.num > o2.num)
				return true;
			else if (o1.num < o2.num)
				return false;
			else
			{
				if (o1.index > o2.index)
					return true;
				else if (o1.index < o2.index)
					return false;
				else
					return true;
			}
		
	}
};

int main()
{

	ifstream fin("fcfs.inp");
	ofstream fout("fcfs.out");

	int n;
	fin >> n;

	vector<fc> v(n);
	vector<bool> pass(n);
	vector<int> endtime(n);
	int total = 0, rest = 0, restf = 0;

	priority_queue<Num, vector<Num>, MyComparator> q;

	//int endtime[n];

	priority_queue<Num, vector<Num>, MyComparator> start;
	//bool pass[n];
	int passn = 0;

	//memset(pass, false, sizeof(pass));

	for (int i = 0; i < n; i++)
	{

		int num;
		deque<int> d;
		fin >> num;
		Num k(i, num);
		start.push(k);

		while (true)
		{
			int end;
			fin >> end;
			if (end == -1)
			{
				d.push_back(end);
				break;
			}
			d.push_back(end);
		}
		v[i].d=d;
		v[i].fin=0;
	}

	int cpu = 0;

	while (passn < n)
	{
		for (int i = 0; i < start.size(); i++)
		{
			Num k = start.top();
			if (total >= k.num)
			{
				q.push(k);
				start.pop();
			}
			else
			{
				break;
			}
		}
		int b = 0;
		if (!q.empty())
		{
			restf = 0;
			Num k = q.top();
			cpu = k.index;
			q.pop();

			b = v[cpu].d.front();
			v[cpu].d.pop_front();

			total += b;
			if (v[cpu].d.front() == -1)
			{
				pass[cpu] = true;
				endtime[cpu] = total;
				passn++;
			}
			else
			{
				v[cpu].fin=2;
			}
		}
		else
		{
			total++;
			rest++;
			restf++;
		}

		for (int i = 0; i < v.size(); i++)
		{
			if (v[i].fin == 1 && !pass[i])
			{
				int a = v[i].d.front();
				v[i].d.pop_front();
				if (b == 0)
				{
					a--;
				}
				else
				{
					a = a - b;
				}
				if (a < 1)
				{
					v[i].fin = 0;
					if (v[i].d.front() == -1)
					{
						pass[i] = true;
						endtime[i] = total + a;
						passn++;
					}
					else
					{
						Num k(i, total + a);
						q.push(k);
					}
				}
				else
				{
					v[i].d.push_front(a);
				}
			}
			if (v[i].fin == 2)
			{
				v[i].fin =1;
			}
		}
	}

	if (restf != 0)
	{
		rest -= restf;
	}

	fout << rest << "\n";

	for (int i = 0; i < n; i++)
	{
		fout << endtime[i] << "\n";
	}
	fout.close();

	return 0;
}

File: 5/fcfs.java
import java.io.*;
import java.util.*;

public class fcfs {

    static class Num {
        public int index;
        public int num;
        public int time;

        Num(int index, int num, int time) {
            this.index = index;
            this.num = num;
            this.time = time;
        }
        void settime(int time){
            this.time = time;
        }
    }

    public static class MyComparator implements Comparator<Num> {
        @Override
        public int compare(Num o1, Num o2) {
            if (o1.time > o2.time)
                return 1;
            else if (o1.time < o2.time)
                return -1;
            else {
                if (o1.num > o2.num)
                    return 1;
                else if (o1.num < o2.num)
                    return -1;
                else {
                    if (o1.index > o2.index)
                        return 1;
                    else if (o1.index < o2.index)
                        return -1;
                    else
                        return 1;
                }
            }
        }
    }
    public static class MComparator implements Comparator<Num> {
        @Override
        public int compare(Num o1, Num o2) {

                if (o1.num > o2.num)
                    return 1;
                else if (o1.num < o2.num)
                    return -1;
                else {
                    if (o1.index > o2.index)
                        return 1;
                    else if (o1.index < o2.index)
                        return -1;
                    else
                        return 1;
                }
            }

    }

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new FileReader("C:\\Users\\Administrator\\Downloads\\OShomwork\\5\\1.txt"));
        BufferedWriter bw = new BufferedWriter(new FileWriter("11.txt"));

        // BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        Vector<Deque<Integer>> v = new Vector<>();

        int total = 0;
        int rest = 0;
        int restf = 0;

        PriorityQueue<Num> q = new PriorityQueue<>(new MyComparator());

        int endtime[] = new int[n];

        PriorityQueue<Num> start = new PriorityQueue<>(new MComparator());

        Boolean[] pass = new Boolean[n];
        int passn = 0;

        for (int i = 0; i < n; i++) {
            pass[i] = false;
        }

        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            Deque<Integer> d = new LinkedList<Integer>();
            int s = Integer.parseInt(st.nextToken());

            while (true) {
                int end = Integer.parseInt(st.nextToken());
                if (end == -1) {
                    d.add(end);
                    d.add(0);
                    break;
                }
                d.add(end);
            }
            Num k = new Num(i, s, d.getFirst());
            start.add(k);
            v.add(d);
        }

        int cpu = 0;

        while (passn < n) {
            for (int i = 0; i < start.size(); i++) {
                Num k = start.peek();
                if (total >= k.num) {
                    q.add(k);
                    start.poll();
                }else{
                    break;
                }
            }
            int b = 0;
            if (q.size() > 0) {
                restf = 0;
                Num k = q.poll();
                cpu = k.index;

                b = v.get(cpu).poll();
                total += b;
                if (v.get(cpu).getFirst() == -1) {
                    pass[cpu] = true;
                    endtime[cpu] = total;
                    passn++;
                } else {
                    v.get(cpu).pollLast();
                    v.get(cpu).addLast(2);
                }
            } else {
                // int min = Integer.MAX_VALUE;

                // for (int i = 0; i < v.size(); i++) {
                //     if (v.get(i).getLast() == 1 && !pass[i]) {
                //         min = v.get(i).getFirst() < min ? v.get(i).getFirst() : min;
                //     }
                // }
                // if (start.size() > 0) {
                //     Num k = start.peek();
                //     if (k.num < total + min) {
                //         q.add(k);
                //         start.poll();
                //         min = k.num - total;
                //     }
                // }
                // total += min;
                // rest += min;
                // restf += min;
                // b = min;
                total++;
                rest++;
                restf++;
            }

            for (int i = 0; i < v.size(); i++) {
                if (v.get(i).getLast() == 1 && !pass[i]) {
                    int a = v.get(i).poll();
                    if (b == 0) {
                        a--;
                    } else {
                        a = a - b;
                    }
                    if (a < 1) {
                        v.get(i).pollLast();
                        v.get(i).addLast(0);
                        if (v.get(i).getFirst() == -1) {
                            pass[i] = true;
                            endtime[i] = total + a;
                            passn++;
                        } else {
                            Num k = new Num(i, total + a, v.get(i).getFirst());
                            q.add(k);
                        }
                    } else {
                        v.get(i).addFirst(a);
                    }
                }
                if (v.get(i).getLast() == 2) {
                    v.get(i).pollLast();
                    v.get(i).addLast(1);
                }
            }
        }

        if (restf != 0) {
            rest -= restf;
        }
        bw.write(rest + "\n");
        for (int i = 0; i < n; i++) {
            bw.write(endtime[i] + "\n");
        }
        bw.close();
    }
}